### PR TITLE
Add ability to alias parameters

### PR DESCRIPTION
**Pull request by kyleknap**
Fixes parameters that shadow reserved keywords in python and inconsistent
parameter names in ec2.

Fixes https://github.com/boto/botocore/issues/772

As to implementation, I followed a similar format (just a tad different) as the CLI so if the alias needed to be in both the CLI and boto3, we could import the dictionary in botocore and modify the ``HIDDEN_ALIAS`` dictionary in the CLI or just copy and paste entities in that dictionary from botocore to the CLI. I went with this path because it required less work and less of a wider impact in the code base then the ``all_member`` route we discussed . Also it would not be difficult for the CLI to pick up these aliases as it already has this functionality.

cc @jamesls @JordonPhillips

### CODE REVIEW

File: tests/unit/test_handlers.py [hunk 1]
```diff
@@ -804,3 +807,77 @@ def test_add_md5_with_bytes_object(self):
         self.assertEqual(
             request_dict['headers']['Content-MD5'],
             'OFj2IjCsPJFfMAxmQxLGPw==')
+
+
+class TestParameterAlias(unittest.TestCase):
```
Comment (jamesls) [inline tests/unit/test_handlers.py]: We should have a test for when a both the alias and the original are entered.

File: tests/unit/test_handlers.py [hunk 2]
```diff
@@ -291,10 +295,9 @@ def test_run_instances_userdata_blob(self):
         # user data.
         user_data = b'\xc7\xa9This is a test'
         b64_user_data = base64.b64encode(user_data).decode('utf-8')
-        event = 'before-parameter-build.ec2.RunInstances'
         params = dict(ImageId='img-12345678',
                       MinCount=1, MaxCount=5, UserData=user_data)
-        self.session.emit(event, params=params)
+        handlers.base64_encode_user_data(params=params)
```
Comment (jamesls) [inline tests/unit/test_handlers.py]: Why did this need to be changed?

File: botocore/handlers.py [hunk 3]
```diff
@@ -628,6 +628,79 @@ def convert_body_to_file_like_object(params, **kwargs):
...
+
+    for original, new_name in aliases.items():
+        event_portion, original_name = original.rsplit('.', 1)
+        parameter_alias = ParameterAlias(original_name, new_name)
+
+        # Add the handlers to the list of handlers.
+        # One handler is to handle when users provide the alias.
+        # The other handler is to update the documentation to show only
+        # the alias.
+        handler_for_parameter_build = (
```
Comment (jamesls) [inline botocore/handlers.py]: Can we rename this?  This isn't actually the handler, this is a tuple of event name to handler.

File: botocore/handlers.py [hunk 4]
```diff
@@ -628,6 +628,79 @@ def convert_body_to_file_like_object(params, **kwargs):
...
+            'before-parameter-build.' + event_portion,
+            parameter_alias.alias_parameter_in_call)
```
Comment (jamesls) [inline botocore/handlers.py]: I think these probably make sense as `REGISTER_FIRST` don't they?  Seems like we want to map these as soon as we possibly can.

**Comment by jamesls** [comment]
Looks good, just had a few questions/comments.

File: tests/unit/test_handlers.py (hunk 2 above)
Comment (kyleknap) [inline tests/unit/test_handlers.py]: By emitting that event, it will call every handler and that fake emission of the event does not include an operation model which is required by the alias handler causing a runtime error. By removing the emission of an event, I feel like that it has become more of a unit test as well as it is decoupled from all of the other handlers that may be registered to this event. Thoughts?

File: tests/unit/test_handlers.py (hunk 1 above)
Comment (kyleknap) [inline tests/unit/test_handlers.py]: What should be the desired behavior? We remove the alias but use the original's value or remove the alias and use the alias's value. Currently because of how its written the alias will clobber the original. Do you think an error should be thrown instead for this scenario saying the parameters provided is ambiguous?

File: botocore/handlers.py (hunk 3 above)
Comment (kyleknap) [inline botocore/handlers.py]: True. Will do.

File: botocore/handlers.py (hunk 4 above)
Comment (kyleknap) [inline botocore/handlers.py]: Yeah we can do that to be safe.

File: tests/unit/test_handlers.py (hunk 2 above)
Comment (JordonPhillips) [inline tests/unit/test_handlers.py]: Agree. Running less unrelated code is better.

File: tests/unit/test_handlers.py (hunk 1 above)
Comment (JordonPhillips) [inline tests/unit/test_handlers.py]: I would say the expected result would be failure on the old without any special ambiguity.

**Comment by JordonPhillips** [comment]
Adding the test @jamesls mentioned would be nice, but otherwise :boat:

**Comment by kyleknap** [comment]
Just added the error case. Should be good to look at.

**Comment by jamesls** [comment]
:shipit:
